keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -687,20 +997,92 @@ We will extend this approach:
-might require a privileged user do the repair.
-
-TODO: We should consider loosening this up and allow admission configuration to
-intercept/guard writes to admission configuration while preventing deadlock -
-Add feature to configure a set of webhooks to intercept other webhooks
-https://github.com/kubernetes/kubernetes/issues/101794.
+Note: In the future we may further loosen this up and allow admission
+configuration to intercept/guard writes to admission configuration while
+preventing deadlock - Add feature to configure a set of webhooks to intercept
+other webhooks https://github.com/kubernetes/kubernetes/issues/101794.
 
 Alternative considered: Each `ValidatingAdmissionPolicy` has a "level", a
 `ValidatingAdmissionPolicy` can match another `ValidatingAdmissionPolicy` of a
 higher level. This could be added later.
 
+#### Reporting violations
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+- Ability to format message strings
+- support for reasons/codes
+
+Example:
+
+```yaml
+# Policy definition
+apiVersion: admissionregistration.k8s.io/v1alpha1
+kind: ValidatingAdmissionPolicy
+metadata:
+  name: "validate-xyz.example.com"
+spec:
+  ...
+  validations:
+    - expression: "self.name.startsWith('xyz-')"
+      level: Deny # the default
+      messageExpression: "self.name + ' must start with \"xyz-\"'"
+      reason: Invalid
+    - expression: "self.name.contains('bad')"
+      level: Warn
+      message: "name contains 'bad' which is discouraged due to ..."
+    - expression: "self.name.contains('suspicious')"
+      level: Audit
+      messageExpression: "self.name + ' contains \"suspicious\"'"
+      auditKey: suspicious-container-name
+```
+
+Equivalent `AdmissionReview`:
+
+```yaml
+request:
+  ...
+response:
+  allowed: false
+  status:
+    #status: Failure
+    message: "nothing-suspicious-or-bad-here must start with \"xyz-\""
+    reason: Invalid
+    code: 422 # Code for "Invalid"
+  auditAnnotations:
+    suspicious-container-name: "nothing-suspicious-or-bad-here contains \"suspicious\""
+  warnings:
+    - "name contains \"bad\" which is discouraged due to ..."
+```
+
+The resulting message would be something like:
+
+```
+configuration validate-config1.example.com of admission policy validate-xyz.example.com denied the request: nothing-suspicious-or-bad-here must start with "xyz-"
+```
+
+Notes:
+
+- using `level` makes it easy to toggle between `Warn` and `Deny`, it also
+  makes `Audit` more consistent with the other outputs (although it does require
+  a key field).
+- `message` is for plain string messages that do not need any formatting
+  (`messageExpression` and `message` would be exclusive)
+- if there are not `Deny` validations, `response.allowed` will automatically be `true`
+
+TODOs:
+
+- What if multiple validations fail?  What if they have different reasons (Invalid + Forbidden)?
+  - Current admission chain short circuits on validating admission rejections? So just return 1st failed?
```

Comment:
This could be annoying for users who then have to discover violations one-by-one

Reply:
I agree. We should be able to support this. I've made a note.